src/nix.rs

```diff
@@ -0,0 +1,338 @@
+                fields.insert(id.into(), rt);
+            }
+            Term::LetPattern(
+                None,
+                destruct::Destruct::Record {
+                    matches: destruct_vec,
+                    open: false,
+                    rest: None,
+                    span,
+                },
+                Term::RecRecord(fields, vec![], Default::default(), None).into(),
+                translate(n.body().unwrap(), file_id),
+            )
+            .into()
+        }
+        ParsedType::With(n) => unimplemented!(),
+
+        ParsedType::Lambda(n) => {
+            let arg = n.arg().unwrap();
+            match arg.kind() {
+                NODE_IDENT => Term::Fun(
+                    arg.to_string().into(),
+                    translate(n.body().unwrap(), file_id),
+                ),
+                NODE_PATTERN => {
+                    use crate::destruct::*;
+                    let pat = rnix::types::Pattern::cast(arg).unwrap();
+                    let at = pat.at().map(|id| id.as_str().into());
+                    // TODO: manage default values:
+                    let matches = pat
+                        .entries()
+                        .map(|e| {
+                            let mv = if let Some(def) = e.default() {
+                                MetaValue {
+                                    value: Some(translate(def, file_id)),
+                                    priority: MergePriority::Default,
+                                    ..Default::default()
+                                }
+                            } else {
+                                Default::default()
+                            };
+                            Match::Simple(e.name().unwrap().as_str().into(), mv)
+                        })
+                        .collect();
+                    let dest = Destruct::Record {
+                        matches,
+                        open: pat.ellipsis(),
+                        rest: None,
+                        span,
+                    };
+                    Term::FunPattern(at, dest, translate(n.body().unwrap(), file_id)).into()
+                }
+                _ => unreachable!(),
+            }
+            .into()
+        }
+
+        ParsedType::Apply(n) => Term::App(
+            translate(n.lambda().unwrap(), file_id),
+            translate(n.value().unwrap(), file_id),
+        )
+        .into(),
+        ParsedType::IfElse(n) => Term::App(
+            Term::App(
+                Term::Op1(UnaryOp::Ite(), translate(n.condition().unwrap(), file_id)).into(),
+                translate(n.body().unwrap(), file_id),
+            )
+            .into(),
+            translate(n.else_body().unwrap(), file_id),
+        )
+        .into(),
+        ParsedType::BinOp(n) => (n, file_id).into(),
+        ParsedType::UnaryOp(n) => (n, file_id).into(),
+        ParsedType::Select(n) => match ParsedType::try_from(n.index().unwrap()).unwrap() {
```

Comment:
Ugh, that's a lot of optionals :stuck_out_tongue:

Reply:
What do you mean? The return type of many functions of `rnix`? Yes in many cases for me it make no sens to return an option...

Reply:
I don't realy know. Why to return an `Option` and not an error.
Here, `try_from` return a `Result` but I didn't manage it for now. But `index()` return an `Option` what will never be `None` if `try_from` don't fail. I think. Anyway if it could, for me it should be concidered as an error (How a record access can have `None` for the accessed field?). It may be done like this for the programaticaly generated trees.

Reply:
> either `rowan` or `rnix` leads to a steady flow of `unwrap` that hurt our little heart of type-safers.

alternatives ...?

imho this should be an incremental parser like tree-sitter or lezer-parser,
so that nickel core can provide the language server